File: publisher-sdk/src/main/java/com/criteo/publisher/model/CdbRequestSlot.java
/*
 *    Copyright 2020 Criteo
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.criteo.publisher.model;

import static java.util.Collections.singletonList;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.criteo.publisher.util.AdUnitType;
import com.google.auto.value.AutoValue;
import com.google.gson.Gson;
import com.google.gson.TypeAdapter;
import com.google.gson.annotations.SerializedName;
import java.util.Collection;
import java.util.List;

@AutoValue
public abstract class CdbRequestSlot {

  public static CdbRequestSlot create(
      @NonNull String impressionId,
      @NonNull String placementId,
      @NonNull AdUnitType adUnitType,
      @NonNull AdSize size
  ) {
    List<String> formattedSizes = singletonList(size.getFormattedSize());

    return new AutoValue_CdbRequestSlot(
        impressionId,
        placementId,
        adUnitType == AdUnitType.CRITEO_CUSTOM_NATIVE ? true : null,
        adUnitType == AdUnitType.CRITEO_INTERSTITIAL ? true : null,
        formattedSizes
    );
  }

  public static TypeAdapter<CdbRequestSlot> typeAdapter(Gson gson) {
    return new AutoValue_CdbRequestSlot.GsonTypeAdapter(gson);
  }

  @NonNull
  @SerializedName("impId")
  public abstract String getImpressionId();

  @NonNull
  @SerializedName("placementId")
  public abstract String getPlacementId();

  // isNative is not accepted by AutoValue because this generates a field called native which is a
  // reserved keyword.
  @Nullable
  @SerializedName("isNative")
  public abstract Boolean isNativeAd();

  @Nullable
  @SerializedName("interstitial")
  public abstract Boolean isInterstitial();

  @NonNull
  @SerializedName("sizes")
  public abstract Collection<String> getSizes();

}

File: publisher-sdk/src/main/java/com/criteo/publisher/model/CdbRequestSlot.kt
/*
 *    Copyright 2020 Criteo
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.criteo.publisher.model

import com.criteo.publisher.annotation.OpenForTesting
import com.criteo.publisher.util.AdUnitType
import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@OpenForTesting
@JsonClass(generateAdapter = true)
data class CdbRequestSlot internal constructor(
    @Json(name = "impId")
    val impressionId: String,
    @Json(name = "placementId")
    val placementId: String,
    @Json(name = "isNative")
    val isNativeAd: Boolean?,
    @Json(name = "interstitial")
    val isInterstitial: Boolean?,
    @Json(name = "rewarded")
    val isRewarded: Boolean?,
    @Json(name = "sizes")
    val sizes: Collection<String>,
    @Json(name = "banner")
    val banner: Banner?
) {

  constructor(
      impressionId: String,
      placementId: String,
      adUnitType: AdUnitType,
      size: AdSize,
      apiFramework: List<ApiFramework>
  ) :
      this(
          impressionId,
          placementId,
          if (adUnitType == AdUnitType.CRITEO_CUSTOM_NATIVE) true else null,
          if (adUnitType == AdUnitType.CRITEO_INTERSTITIAL) true else null,
          if (adUnitType == AdUnitType.CRITEO_REWARDED) true else null,
          listOf(size.formattedSize),
          apiFramework.takeIf { it.isNotEmpty() }?.run { Banner(api = this.map { it.code }) }
      )
}

@JsonClass(generateAdapter = true)
data class Banner internal constructor(
    @Json(name = "api")
    val api: List<Int>
)
